daemon/logger/loggerutils/sharedtemp.go

```diff
@@ -0,0 +1,225 @@
+package loggerutils // import "github.com/docker/docker/daemon/logger/loggerutils"
+
+import (
+	"io"
+	"io/fs"
+	"os"
+	"runtime"
+	"sync"
+)
+
+type fileConvertFn func(dst io.WriteSeeker, src io.ReadSeeker) error
+
+type stfID uint64
+
+// sharedTempFileConverter converts files using a user-supplied function and
+// writes the results to temporary files which are automatically cleaned up on
+// close. If another request is made to convert the same file, the conversion
+// result and temporary file are reused if they have not yet been cleaned up.
+//
+// A file is considered the same as another file using the os.SameFile function,
+// which compares file identity (e.g. device and inode numbers on Linux) and is
+// robust to file renames. Input files are assumed to be immutable; no attempt
+// is made to ascertain whether the file contents have changed between requests.
+//
+// One file descriptor is used per source file, irrespective of the number of
+// concurrent readers of the converted contents.
+type sharedTempFileConverter struct {
+	C       fileConvertFn
+	TempDir string
+
+	st   chan stfcState
+	init sync.Once
+}
+
+type stfcState struct {
+	fl     map[stfID]sharedTempFile
+	nextID stfID
+}
+
+type sharedTempFile struct {
+	src  os.FileInfo // Info about the source file for path-independent identification with os.SameFile.
+	fd   *os.File
+	size int64
+	ref  int                       // Reference count of open readers on the temporary file.
+	wait []chan<- stfConvertResult // Wait list for the conversion to complete.
+}
+
+type stfConvertResult struct {
+	fr  *sharedFileReader
+	err error
+}
+
+// Do returns a reader for the contents of f as converted by the c.C function.
+// It is the caller's responsibility to close the returned reader.
+//
+// This function is safe for concurrent use by multiple goroutines.
+func (c *sharedTempFileConverter) Do(f *os.File) (*sharedFileReader, error) {
+	c.init.Do(func() {
+		c.st = make(chan stfcState, 1)
+		c.st <- stfcState{fl: make(map[stfID]sharedTempFile)}
+	})
```

Comment:
Could maybe introduce `NewSharedTempFileConverter(fileConvertFn)` to make it clearer the required attributes and initializer logic? Using `sync.Once` to workaround lack of constructors in go sounds a bit hack-ish to me :P

Reply:
Using `sync.Once` here is a minor optimization to lazily initialize state (and allocate memory) on first use as the converter is not used unless there is a compressed rotated log file to read. But it's probably such a trivial amount of improvement—and a premature optimization—that it's not worth the readability hit.